tests/v2/validation/workloads/deployment_test.go

```diff
@@ -0,0 +1,308 @@
+//go:build (validation || infra.any || cluster.any || sanity) && !stress && !extended
+
+package workloads
+
+import (
+	"testing"
+
+	projectsapi "github.com/rancher/rancher/tests/v2/actions/projects"
+	deployment "github.com/rancher/rancher/tests/v2/actions/workloads/deployment"
+	"github.com/rancher/rancher/tests/v2/actions/workloads/pods"
+	"github.com/rancher/shepherd/clients/rancher"
+	management "github.com/rancher/shepherd/clients/rancher/generated/management/v3"
+	"github.com/rancher/shepherd/extensions/charts"
+	"github.com/rancher/shepherd/extensions/clusters"
+	"github.com/rancher/shepherd/extensions/workloads"
+	namegen "github.com/rancher/shepherd/pkg/namegenerator"
+	"github.com/rancher/shepherd/pkg/session"
+	log "github.com/sirupsen/logrus"
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type DeploymentTestSuite struct {
+	suite.Suite
+	client  *rancher.Client
+	session *session.Session
+	cluster *management.Cluster
+}
+
+func (d *DeploymentTestSuite) TearDownSuite() {
+	d.session.Cleanup()
+}
+
+func (d *DeploymentTestSuite) SetupSuite() {
+	d.session = session.NewSession()
+
+	client, err := rancher.NewClient("", d.session)
+	require.NoError(d.T(), err)
+
+	d.client = client
+
+	log.Info("Getting cluster name from the config file and append cluster details in connection")
+	clusterName := client.RancherConfig.ClusterName
+	require.NotEmptyf(d.T(), clusterName, "Cluster name to install should be set")
+
+	clusterID, err := clusters.GetClusterIDByName(d.client, clusterName)
+	require.NoError(d.T(), err, "Error getting cluster ID")
+
+	d.cluster, err = d.client.Management.Cluster.ByID(clusterID)
+	require.NoError(d.T(), err)
+}
+
+func (d *DeploymentTestSuite) TestDeploymentSideKick() {
+	subSession := d.session.NewSession()
+	defer subSession.Cleanup()
+
+	log.Info("Creating new project and namespace")
+	_, namespace, err := projectsapi.CreateProjectAndNamespace(d.client, d.cluster.ID)
+	require.NoError(d.T(), err)
+
+	log.Info("Creating new deployment")
+	createdDeployment, err := deployment.CreateDeployment(d.client, d.cluster.ID, namespace.Name, 1, "", "", false, false)
+	require.NoError(d.T(), err)
+
+	log.Info("Waiting for all pods to be running")
+	err = pods.WatchAndWaitPodContainerRunning(d.client, d.cluster.ID, namespace.Name, createdDeployment)
+	require.NoError(d.T(), err)
+
+	containerName := namegen.AppendRandomString("update-test-container")
+	newContainerTemplate := workloads.NewContainer(containerName,
+		redisImageName,
+		corev1.PullAlways,
+		[]corev1.VolumeMount{},
+		[]corev1.EnvFromSource{},
+		nil,
+		nil,
+		nil,
+	)
```

Comment:
One thing I'm noticing is that this function is called repeatedly with the same arguments. I wonder if it can be less repeated by creating another function with less parameters?

Reply:
Yes, I can

Reply:
Maybe like this?
```
func newRedisContainerTemplate(containerName string) *corev1.Container {
    return workloads.NewContainer(containerName,
        redisImageName,
        corev1.PullAlways,
        []corev1.VolumeMount{},
        []corev1.EnvFromSource{},
        nil,
        nil,
        nil,
    )
}
```

Reply:
Maybe it could take the image name as a parameter

Reply:
Based on our previous design discussion, we want to avoid creating multiple layers of abstraction. Since workloads.NewContainer is already an abstraction, creating another one could cause confusion. Instead, we can repeat the same steps directly in the tests to improve readability.

Reply:
So, could we keep it the same way? @floatingman @anupama2501

Reply:
yes IMHO.

Reply:
It is a maintenance hardship since the parameters are the same across the three functions, but it's really not that important.